Improve Integration Tests with switch to Chrome Driver

jogep opened the pull request.

jogep committed:
Improve Integration Tests with switch to Chrome Driver

* Setup ChromeDriver for Integration tests
* More realistic tests
* Fix TabbedpanelTagIT, AccordionTagIT and SliderTagIT
* Disable TreeTagIT (needs dedicated investigation)

https://github.com/struts-community-plugins/struts2-jquery/issues/709

jogep committed:
Add missing profile for Chrome Driver

lukaszlenart commented:
I tested this PR locally on JDK17 and it passes - I'm a bit lost :\

jogep committed:
Add some debug output to workflow for deep dive

jogep committed:
Remove unnecessary web.xml override and extend the usage of WebDriverFactory for chrome driver instantiation

jogep committed:
Increase testing timeout and don't use static for WebDriver

jogep committed:
Add additional wait time check if Struts2 jQuery file is initialized + disable parallel execution and increase timeout

jogep committed:
Change the way how ChromeDriver is handled in CI vs Local development

jogep committed:
Change workflow step to Setup Chrome and ChromeDriver

jogep committed:
Improve mvn start command

jogep committed:
Add some debug logging for dive deep

jogep committed:
Add some ChromeDriver options

jogep committed:
Add additional debug steps and more fine grained build order

jogep committed:
Merge and simplify integration test step

jogep committed:
Remove duplicate execution of integration tests

jogep committed:
Improve logging and jetty configuration

jogep committed:
Add some debug logging to analyze why static resources are not loaded in CI environment

jogep committed:
Add dedicated jetty logging config

jogep committed:
Correct intention

lukaszlenart commented:
Maybe use a different port than 8080?

jogep committed:
Increase looging output lines

jogep commented:
@lukaszlenart I don't think the port is the problem as jetty is running and struts2 actions are working fine:

```
🔍 Step 4: Analyzing Struts2 behavior...
Testing a working Struts2 action page:
Found working action: /regular/div/ajax-div.action
  Working action HTTP: 200
  ✅ Struts2 actions are working
```

For some resons the static resources are not provided and I got a 404 in the CI environment while it works locally.

```

Testing: http://localhost:8080/static/js/struts2/jquery.struts2.min.js?s2j=6.0.2-SNAPSHOT
  HTTP: 404 | Size: 472 bytes | Time: 0.004534s
  ❌ FAILED
  Response body: <html>
<head>
<meta http-equiv="Content-Type" content="text/html;charset=ISO-8859-1"/>
<title>Error 404 Not Found</title>
</head>
<body><h2>HTTP ERROR 404 Not Found</h2>
<table>
<tr><th>URI:</th><td>/
  Response headers:
HTTP/1.1 404 Not Found
Date: Sun, 01 Jun 2025 20:55:33 GMT
Cache-Control: must-revalidate,no-cache,no-store
Content-Type: text/html;charset=iso-8859-1
Content-Length: 472
```

lukaszlenart commented:
I made a few changes locally, first used
```
		<dependency>
			<groupId>org.apache.logging.log4j</groupId>
			<artifactId>log4j-slf4j2-impl</artifactId>
		</dependency>
```
instead of 
```
		<dependency>
			<groupId>org.apache.logging.log4j</groupId>
			<artifactId>log4j-slf4j-impl</artifactId>
		</dependency>
```
to solve warning with missing SL4j bindings.


Then I added 
```
		<dependency>
			<groupId>jakarta.servlet.jsp</groupId>
			<artifactId>jakarta.servlet.jsp-api</artifactId>
		</dependency>
```

and when I'm starting `mvn verify -P ci-chrome` I get a few errors, eg.:

```
Cannot find function escape in object [object CSS]. (https://code.jquery.com/jquery-3.7.1.min.js#2)
```

jogep committed:
Add config browser and correct log4j-slf4j2-impl dependency

jogep commented:
Where did you add the jakarta.servlet.jsp-api and why? If you get this error message the static is not served as well in your case. I don't get the error in my environment and statics are served fine in last version of this branch.

lukaszlenart commented:
I've added it into integration-tests pom.xml just next to `<artifactId>jakarta.servlet-api</artifactId>` and because I assumed it should be defined there same as `jakarta.servlet-api`. And the statics are served, yet it looks like versions are mismatched.

jogep committed:
Add missing jakarta.servlet.jsp-api dependency to integration tests

jogep committed:
Print full jetty-detailed.log for stack trace

jogep commented:
I added your suggested changes, it still fails. I changed the “Debug Struts2 Static Resource Serving” workflow step to provide the full log but with no additional details. Only one stack trace, but this is the single / root request to check if jetty is up and running. Don't really understand why the static is not provided by the Struts2 filter while the actions itself are working fine. Also because the statics are served on my dev environment with same maven command without any issues.

lukaszlenart commented:
But not all the tests fail

```
Tests run: 173, Failures: 0, Errors: 126, Skipped: 5
```

Maybe disable all the failing tests and enable just one of them?

lukaszlenart commented:
or disable Maven cache, maybe one of the downloaded artifacts got broken

jogep committed:
Add default to prevent using jetty using webdefault.xml

jogep commented:
@lukaszlenart are you using the latest version? Locally I don't have any errors and the build returns with SUCCESS.

```
[WARNING] Tests run: 173, Failures: 0, Errors: 0, Skipped: 5
[INFO] 
[INFO] 
[INFO] --- jetty:11.0.25:stop (stop-jetty) @ struts2-jquery-integration-tests ---
[INFO] Stopping jetty
[INFO] Stopped Server@26d7cb0d{STOPPING}[11.0.25,sto=0]
[INFO] Stopped ServerConnector@192aeae2{HTTP/1.1, (http/1.1)}{0.0.0.0:8080}
[INFO] 
[INFO] --- failsafe:3.5.3:verify (default) @ struts2-jquery-integration-tests ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
```


All static files referenced in the generated html are loaded locally, But in the CI environments you see this logs:

```
Error: 11.801] [ERROR] [com.jgeppert.jquery.AbstractJQueryTest] - ❌ Timeout waiting for page load. Debug info:
[15:56:11.805] [WARN] [com.jgeppert.jquery.AbstractJQueryTest] - 🔍 Current URL: http://localhost:8080/regular/radio/remotelist.action
[15:56:11.810] [WARN] [com.jgeppert.jquery.AbstractJQueryTest] - 🔍 Browser console logs:
[15:56:11.811] [WARN] [com.jgeppert.jquery.AbstractJQueryTest] -   SEVERE: http://localhost:8080/static/js/base/version.min.js - Failed to load resource: the server responded with a status of 404 (Not Found)
[15:56:11.811] [WARN] [com.jgeppert.jquery.AbstractJQueryTest] -   SEVERE: http://localhost:8080/static/js/struts2/jquery.struts2.min.js?s2j=6.0.2-SNAPSHOT - Failed to load resource: the server responded with a status of 404 (Not Found)
[15:56:11.811] [WARN] [com.jgeppert.jquery.AbstractJQueryTest] -   SEVERE: http://localhost:8080/static/js/plugins/jquery.subscribe.min.js?s2j=6.0.2-SNAPSHOT - Failed to load resource: the server responded with a status of 404 (Not Found)
[15:56:11.811] [WARN] [com.jgeppert.jquery.AbstractJQueryTest] -   WARNING: http://localhost:8080/static/js/base/jquery-3.7.1.min.js 1:28647 "jQuery.Deferred exception: Cannot set properties of undefined (setting 'version')" "TypeError: Cannot set properties of undefined (setting 'version')\n    at HTMLDocument.\u003Canonymous> (http://localhost:8080/regular/radio/remotelist.action:23:39)\n    at e (http://localhost:8080/static/js/base/jquery-3.7.1.min.js:2:27028)\n    at t (http://localhost:8080/static/js/base/jquery-3.7.1.min.js:2:27330)" undefined
[15:56:11.811] [WARN] [com.jgeppert.jquery.AbstractJQueryTest] -   WARNING: http://localhost:8080/static/js/base/jquery-3.7.1.min.js 1:28647 "jQuery.Deferred exception: Cannot read properties of undefined (reading 'bind')" "TypeError: Cannot read properties of undefined (reading 'bind')\n    at HTMLDocument.\u003Canonymous> (http://localhost:8080/regular/radio/remotelist.action:60:26)\n    at e (http://localhost:8080/static/js/base/jquery-3.7.1.min.js:2:27028)\n    at t (http://localhost:8080/static/js/base/jquery-3.7.1.min.js:2:27330)" undefined
Error: 11.799] [ERROR] [com.jgeppert.jquery.AbstractJQueryTest] - ❌ Timeout waiting for page load. Debug info:
```

Tried different approaches but don't get the serving of the statics by Struts2 working.

jogep committed:
Add webdefault.xml to control default behavior of jetty

jogep committed:
Add new Debug Actual Request Flow CI workflow step

jogep committed:
Completely remove default servlet references

jogep committed:
Remove debug step as servlet mapping is not exiting anymore

lukaszlenart commented:
It fails for me when using
```
mvn verify -P ci-chrome
```

what version of Java do you use? I use:
```
dev-3:[17](sml:) ~/P/S/s/struts2-jquery-integration-tests (feature/it-chrome-driver)> java -version
openjdk version "17.0.14" 2025-01-21
OpenJDK Runtime Environment Temurin-17.0.14+7 (build 17.0.14+7)
OpenJDK 64-Bit Server VM Temurin-17.0.14+7 (build 17.0.14+7, mixed mode, sharing)
```

I noticed GH Actions are running 17.05

jogep commented:
@lukaszlenart can you try the same command like defined in the workflow?

```
        mvn -pl struts2-jquery-integration-tests \
        -Dmaven.javadoc.skip=true \
        -Dmaven.source.skip=true \
        -Djetty.scan=0 \
        -Djava.util.logging.config.file=src/test/resources/jetty-logging.properties \
        -P ci-chrome \
        -Dfailsafe.webdriver.name=ChromeDriver \
        verify
```

Do you think the Java version could be a cause for the different handling of static files?
I use locally the Corretto distribution, I could give it a try but would strange.

Same error with Corretto:

```
Run java -version
openjdk version "1[7](https://github.com/struts-community-plugins/struts2-jquery/actions/runs/15400464697/job/43331560513?pr=714#step:6:8).0.15" 2025-04-15 LTS
OpenJDK Runtime Environment Corretto-17.0.15.6.1 (build 17.0.15+6-LTS)
OpenJDK 64-Bit Server VM Corretto-17.0.15.6.1 (build 17.0.15+6-LTS, mixed mode, sharing)
```

jogep committed:
Try corretto distribution for Java setup

jogep committed:
Try newer version of maven jetty plugin

lukaszlenart commented:
With the above command tests passed, yet when I removed `-pl` and started command directly in the folder one of the tests failed. Can this be the case?

jogep committed:
Replace Thread.sleep with wait command to fix jetty stop problem when executing tests within the package

jogep commented:
@lukaszlenart you are right with the package the same command fails what is somehow strange. I pushed a fix for the issue and locally it works. However, I don't think the issue is not related to the "not serving statics" in CI environment. When you take a look at the debug output in the action runs you see that the statics like jQuery files return with a 404 error.

jogep committed:
Add some jetty settings to improve behavior

lukaszlenart commented:
I thought maybe hosting static files depends on the root folder used to start the process, so starting `mvn verify` from the integration-tests folder can have different behaviour.

jogep committed:
Add debug step to check struts constants in CI env

jogep committed:
Change extractWAR from true to false and add parentLoaderPriority in the Jetty configuration

jogep committed:
Add ResourceDebugServlet for resource issue debugging in CI step

jogep committed:
Remove duplicate conflicting registration of debug servlet

jogep committed:
Add the missing build plugins part in pom.xml files

jogep committed:
Bring back/Add missing plugin section in poms to generate minified files

jogep commented:
@jogep finally I found the root cause! Some stupid (me) has removed, and looks like in some plugins it was missing, the build plugins section. With that the minification has not worked properly. Will do some cleanup before it is ready to merge.

jogep committed:
Clean up verbose logging and debug, unify jetty plugin version usage

lukaszlenart reviewed (approved):
💯
